rebasehelper/build_log_hooks/files_build_log_hook.py

```diff
@@ -0,0 +1,257 @@
+        """Adds files found in sources to the SPEC file."""
+        macros = [m for m in rebase_spec_file.macros if m['name'] in MacroHelper.MACROS_WHITELIST]
+        macros = MacroHelper.expand_macros(macros)
+        # ensure maximal greediness
+        macros.sort(key=lambda k: len(k['value']), reverse=True)
+
+        result = collections.defaultdict(lambda: collections.defaultdict(list))
+
+        for file in files:
+            section = cls._get_best_matching_files_section(rebase_spec_file, file)
+            substituted_path = MacroHelper.substitute_path_with_macros(file, macros)
+            rebase_spec_file.spec_content.sections[section].insert(0, substituted_path)
+            result['added'][section].append(substituted_path)
+            logger.info('Added %s to %s section', substituted_path, section)
+
+        return result
+
+    @classmethod
+    def _correct_deleted_files(cls, rebase_spec_file, files):
+        """Removes files missing in sources from the SPEC file."""
+        result = collections.defaultdict(lambda: collections.defaultdict(list))
+        for sec_name, sec_content in six.iteritems(rebase_spec_file.spec_content.sections):
+            if sec_name.startswith('%files'):
+                i = 0
+                while i < len(sec_content):
+                    original_line = sec_content[i].split()
+                    if not original_line:
+                        i += 1
+                        continue
+                    split_line = original_line.copy()
+                    directives = []
+                    prepend_macro = None
+                    for element in reversed(split_line):
+                        if element in cls.FILES_DIRECTIVES:
+                            if cls.FILES_DIRECTIVES[element]:
+                                prepend_macro = cls.FILES_DIRECTIVES[element]
+                            directives.insert(0, element)
+                            split_line.remove(element)
+
+                    if prepend_macro:
+                        split_line = [os.path.join(prepend_macro, file) for file in split_line]
+                    split_line = [MacroHelper.expand(f) for f in split_line]
+
+                    j = 0
+                    while j < len(split_line) and files:
+                        file = split_line[j]
+                        for deleted_file in reversed(files):
+                            if not fnmatch.fnmatch(deleted_file, file):
+                                j += 1
+                                continue
+
+                            original_file = original_line[len(directives) + j]
+
+                            del split_line[j]
+                            del original_line[len(directives) + j]
+                            files.remove(deleted_file)
+                            result['removed'][sec_name].append(original_file)
+                            logger.info('Removed %s from %s section', original_file, sec_name)
```

Comment:
I may be missing something, but shouldn't you increment `j` here?

Reply:
If I incremented `j`, wouldn't I skip a file? `j` is the index of the file currently inspected, if I delete element on index `j` then the next file which hasn't been inspected yet is still on `j`, isn't it?

Reply:
Of course, that's what I was missing :-)